sdk/servicebus/Azure.Messaging.ServiceBus/src/Plugins/ServiceBusPlugin.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.Messaging.ServiceBus.Plugins
+{
+    /// <summary>
+    /// This class provides methods that can be overridden to manipulate messages for custom plugin functionality.
+    /// </summary>
+    public abstract class ServiceBusPlugin
+    {
+        /// <summary>
+        /// Gets the name of the <see cref="ServiceBusPlugin"/>.
+        /// </summary>
+        public abstract string Name { get; }
+
+        /// <summary>
+        /// Determines whether or an exception in the plugin should prevent a send or receive operation.
+        /// </summary>
+        public virtual bool ShouldContinueOnException => false;
+
+        /// <summary>
+        /// This operation is called before a message is sent and can be
+        /// overridden to alter the properties of an outgoing message.
+        /// </summary>
+        /// <param name="message">The <see cref="ServiceBusMessage"/> to be modified by the plugin.</param>
+        public virtual Task BeforeMessageSend(ServiceBusMessage message)
+        {
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// This operation is called after a message is received, but before it is returned to the <see cref="ServiceBusReceiver"/>.
+        /// It can be overriden to alter the properties of an
+        /// incoming message.
+        /// </summary>
+        /// <param name="message">The <see cref="ServiceBusReceivedMessage"/> to be modified by the plugin.</param>
+        public virtual Task AfterMessageReceive(ServiceBusReceivedMessage message)
+        {
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Set the <see cref="ServiceBusReceivedMessage.Body"/>.
+        /// </summary>
+        /// <param name="message">The message to modify.</param>
+        /// <param name="body">The body to set on the message. This will overwrite any existing body.</param>
+        protected static void SetBody(ServiceBusReceivedMessage message, BinaryData body)
```

Comment:
Did you mean `static` for these?   It seems these would likely be more discoverable by plugin authors if they were on the instance.

Reply:
Yeah these need to be static otherwise FxCop complains as it doesn't access instance members. I can work around this if needed I guess.

Reply:
I don't feel strongly, but this is a case where I don't think FxCop has the proper context around the semantics and is offering advice that isn't ideal.   As a plugin author exploring the abstract class that I'm extending,  I'm more likely to type `this.` and see what pops up than `ServiceBusPlugin.`